creusot-contracts/src/std/iter/filter.rs

```diff
@@ -0,0 +1,129 @@
+    #[predicate(prophetic)]
+    #[open(self)]
+    fn invariant(self) -> bool {
+        pearlite! {
+            // trivial precondition: simplification for sake of proof complexity
+            forall<f : F, i : &I::Item> f.precondition((i,)) &&
+            // immutable state: simplification for sake of proof complexity
+            (forall<f : F, g : F> f.unnest(g) ==> f == g) &&
+            // plain-ness: this is a simplification as well, but a fairly major one for very minor loss of expressivity. In short all it says
+            // is that we cannot `a == b` two mutable borrows.
+            (forall<f : &mut F, g : &mut F, i :_, b :_> *f == *g && ^f == ^g ==> f.postcondition_mut((i,), b) == g.postcondition_mut((i,), b)) &&
+            // precision of postcondition. In some sense this is not *necessary*, but without this we cannot prove that we return *all* elements
+            // for all elements where the predicate evaluated to true, since we don't actually have access to the closure's return value directly,
+            // only what the postcondition says about it.
+            (forall<f : &mut F, i : _> f.postcondition_mut((i,), true) != f.postcondition_mut((i,), false))
+        }
+    }
+}
+
+/// Asserts that `f` has no precondition: any closure state can be called with any input value
+/// In a future release this restriction may be lifted or weakened
+#[open]
+#[predicate]
+pub fn no_precondition<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : F, i : A> f.precondition((i,)) }
+}
+
+/// Asserts that the captures of `f` are used immutably
+/// In a future release this restriction may be lifted or weakened
+#[open]
+#[predicate]
+pub fn immutable<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : F, g : F> f.unnest(g) ==> f == g }
+}
+
+/// The postcondition of `f` does not depend on prophecies storied in its in environment.
+/// In practice this means that the postcondition is not testing the equality of a mutable borrow
+#[open]
+#[predicate(prophetic)]
+pub fn plain<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : &mut F, g : &mut F, i :_, b :_> *f == *g && ^f == ^g ==> f.postcondition_mut((i,), b) == g.postcondition_mut((i,), b) }
+}
+
+/// Asserts that the postcondition of `f` is *precise*: that there are never two possible values matching the postcondition
+#[open]
+#[predicate]
+pub fn precise<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : &mut F, i : _> f.postcondition_mut((i,), true) != f.postcondition_mut((i,), false) }
+}
+
+impl<I, F> Iterator for Filter<I, F>
+where
+    I: Iterator,
+    F: FnMut(&I::Item) -> bool,
+{
+    #[open]
+    #[predicate(prophetic)]
+    fn completed(&mut self) -> bool {
+        pearlite! {
+        (exists<inner : &mut _> inv(inner) && *inner == self.iter() && ^inner == (^self).iter()
+            && inner.completed()) && (*self).func() == (^self).func() }
```

Comment:
I don't see why this is correct. When `Filter::next` returns, the underlying iterator may have performed several steps.

Reply:
That's not forbidden?

Reply:
lol nope, i can prove false trivially, yet no proof fails! unsoundness time!

Reply:
Why would it be forbidden? If `Filter` does not find an item that satisfy the predicate, then it returns `None`, even if several items have been produced by the underlying iterator.

Reply:
agreed, i meant that it wasn't supposed to be forbidden, but now i notice that I never exercised that spec due to an unsoundness in our encoding.